File: tps/01/Palindromo.c
#include <stdio.h>
#include <stdbool.h>
#include <locale.h>
#include <wchar.h>

// funcao que verifica se é um palindromo, recebendo uma letra inicial e uma final
bool isPalindrum(const wchar_t *str, int inicio, int fim) {
	// Se chegar no meio, ele retorna que é um palindromo
	if (inicio >= fim) {
		return true;
	}

	// Compara se a letra inicial e final é diferente, se sim então não é um palindromo
	if (str[inicio] != str[fim]) {
		return false;
    	}

	// Chama a função novamente, passando para a proxima letra
	return isPalindrum(str, inicio + 1, fim - 1);
}

int main() {
	setlocale(LC_ALL, ""); // codificação UTF-8
	wchar_t text[100];
    
	while (true) {
		// Pega o que o usuário digitou
		fgetws(text, 100, stdin); 
		
		// Retira a quebra de linha do final da palavra
		text[wcscspn(text, L"\n")] = L'\0';

		// Compara se o texto digitado é igual a FIM, se sim então quebra o loop
		if (wcscmp(text,L"FIM") == 0) {
			break;
		}
	 
		// Pega o tamanho da string, pra saber qual a letra final
   		int size = wcslen(text);

   		// Roda a função de palindromo, passando a posição da letra inicial e da final
    		if (size > 0 && isPalindrum(text, 0,  size - 1)) {
        		wprintf(L"SIM\n");
    		} else {
        		wprintf(L"NAO\n");
    		}
    	}

    return 0;
}

File: tps/01/Palindromo.java
import java.util.Scanner;

public class Palindromo {
	public static void main(String[] args) {
		// Cria um scanner para recuperar os digitos do usuário
		Scanner scan = new Scanner(System.in);

		while (true) {
			String s = scan.nextLine(); // Le o proximo texto digitado

			if (isFim(s)) {
				break;
			}
			
			if (isPalindrum(s)) { // Roda a função e verifica se é palindromo ou não
				MyIO.println("SIM");
			} else {
				MyIO.println("NAO");
			}
		}
	}

	// Uma função que verifica se a string é igual a FIM, que vai ser usada para parar o loop principal
	public static boolean isFim(String s) {
		if (s.length() == 3) {
			if (s.charAt(0) == 'F' && s.charAt(1) == 'I' && s.charAt(2) == 'M') {
				return true;
			}
		}

		return false;
	}

	public static boolean isPalindrum(String s) {
		// Passa por cada letra do inicio ao final comparando se são iguais
		// Quando o inicio se torna igual o final, então siginifica que chegou no meio da palavra e quebra o loop
		for (int i = 0, j = s.length() - 1; i < j; i++, j--) {
			if (s.charAt(i) != s.charAt(j)) { // Verifica se a letra no inicio e no final são iguais, senão quebra o loop e retorna não é um palindromo
				return false;
			}
		}
		
		// Caso todas as letras sejam iguais então é palindromo
		return true;
	}
}
